### ISSUE TITLE
ZAP fails to import recursive OpenAPI definitions

### ISSUE THREAD

**Issue by cedricbu** [zaproxy/zaproxy#7346]
### Describe the bug

When attempting to import an OAS definition which happens to have recursions, it causes a stack overflow in ZAP, and the import fails.

### Steps to reproduce the behavior

1. Create an OAS, that contains some recursions. e.g.:

```json
1 {
   2     "components": {
[...]
 307         "schemas": {
[...]
1730             "SystemProfileNestedObject": {
1731                 "additionalProperties": {
1732                     "oneOf": [
1733                         {
1734                             "$ref":
"#/components/schemas/SystemProfileNestedObject"
1738                         },
[...]
```

2. Load it in ZAP ( "import" -> "Import an OpenAPI definition..."
3. Select the API you have created earlier, and click 'Import'
4. ... nothing happens

### Expected behavior

Recursion should not prevent the OAS to be imported. 

### Software versions

ZAP 2.11.1

### Errors from the zap.log file

2022-06-13 16:30:01,587 [ZAP-ProxyThread-78] ERROR UncaughtExceptionLogger - Exception in thread "ZAP-ProxyThread-78"
java.lang.StackOverflowError: null
        at io.swagger.v3.parser.util.ResolverFully.resolveSchema(ResolverFully.java:413) ~[?:?]
        at io.swagger.v3.parser.util.ResolverFully.resolveSchema(ResolverFully.java:350) ~[?:?]
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655) ~[?:?]
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[?:?]
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) ~[?:?]
        at io.swagger.v3.parser.util.ResolverFully.resolveSchema(ResolverFully.java:413) ~[?:?]
        at io.swagger.v3.parser.util.ResolverFully.resolveSchema(ResolverFully.java:350) ~[?:?]
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655) ~[?:?]
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[?:?]
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) ~[?:?]
        at io.swagger.v3.parser.util.ResolverFully.resolveSchema(ResolverFully.java:413) ~[?:?]
        at io.swagger.v3.parser.util.ResolverFully.resolveSchema(ResolverFully.java:350) ~[?:?]
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655) ~[?:?]
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[?:?]
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) ~[?:?]
        at io.swagger.v3.parser.util.ResolverFully.resolveSchema(ResolverFully.java:413) ~[?:?]
        at io.swagger.v3.parser.util.ResolverFully.resolveSchema(ResolverFully.java:350) ~[?:?]
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655) ~[?:?]
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[?:?]
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) ~[?:?]
        at io.swagger.v3.parser.util.ResolverFully.resolveSchema(ResolverFully.java:413) ~[?:?]
        at io.swagger.v3.parser.util.ResolverFully.resolveSchema(ResolverFully.java:350) ~[?:?]
[...]

### Would you like to help fix this issue?

- [X] Yes

**Comment by thc202** [comment]
Based on the stacktrace this seems to be a problem with the Swagger library.

**Comment by cedricbu** [comment]
Reported to Swagger : https://github.com/swagger-api/swagger-parser/issues/1751

**Comment by thc202** [comment]
Thank you.

**Comment by cedricbu** [comment]
Although I tried to use the https://editor.swagger.io/ against the problematic OAS, and it doesn't seem to any cause problem there.
* I am not able to find which version of swagger-parser is being used by ZAP. Maybe ZAP uses an old version ? (there were known recursion bugs in the past, per swagger-parser git history)
* Or could it be that ZAP doesn't use the library correctly ?

I simplified as much as I could the OAS, down to the following (which causes a stackoverflow in ZAP, but works on https://editor.swagger.io/) :


```yaml
---
components:
  parameters:
    p_one:
      in: query
      name: "a"
      schema:
        "$ref": "#/components/schemas/NestedObject"
      style: deepObject
    p_two:
      in: query
      name: b
      schema:
        "$ref": "#/components/schemas/NestedObject"
      style: deepObject
  schemas:
    NestedObject:
      additionalProperties:
        oneOf:
        - "$ref": "#/components/schemas/NestedObject"
        - not:
            type: object
      type: object
info:
  title: Schema component Recursion
  version: 1.0.0
openapi: 3.0.0
paths:
  "/first":
    delete:
      responses:
        200: 
          description: ok
      parameters:
      - "$ref": "#/components/parameters/p_two"
    get:
      responses:
        200: 
          description: ok
      parameters:
      - "$ref": "#/components/parameters/p_two"
      - "$ref": "#/components/parameters/p_one"
  "/second":
    get:
      responses:
        200: 
          description: ok
      parameters:
      - "$ref": "#/components/parameters/p_one"
servers:
- url: "/api"
```

**Comment by thc202** [comment]
Yes, I did think that could have been fixed (there was a new release yesterday) and read the latest code, the lines that lead to the overflow still match. I did not actually try running with the latest version though will do that now. Worth noting that we were not using the latest version (prior the yesterday's release) because it introduced other issues.

1) The OpenAPI add-on is in zap-extensions repo and its dependencies declared here: https://github.com/zaproxy/zap-extensions/blob/204a09d8d94c4d0acd0d659494cd9e20500121cd/addOns/openapi/openapi.gradle.kts
2) Even if ZAP is doing something wrong the library should not end up in a stack overflow.

It might be that the Swagger Editor does not try to resolve the definition.

**Comment by thc202** [comment]
The stack overflow still happens with the latest version (tested in isolation and with ZAP).

**Comment by thc202** [comment]
This has been fixed upstream and already released in the OpenAPI add-on.
